File: C++/ps_baekjoon/2157/main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int INF = numeric_limits<int>::max() / 2;
int n, m, k, res = 0;
vector<vector<pair<int, int>>> graph;
vector<vector<int>> dp;

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> m >> k;
    graph.resize(n + 1);
    dp = vector<vector<int>>(m, vector<int>(n + 1, -INF));
    dp[0][1] = 0;
    while (k--)
    {
        int a, b, c;
        cin >> a >> b >> c;
        if (!(a < b))
            continue;
        graph[a].emplace_back(b, c);
    }
    for (int from = 0; from <= n; ++from)
        for (const auto& edge : graph[from])
        {
            auto [to, cost] = edge;
            for (int cnt = 1; cnt < m; ++cnt)
                dp[cnt][to] = max(dp[cnt][to], dp[cnt - 1][from] + cost);
        }
    for (int i = 1; i < m; ++i)
        res = max(res, dp[i][n]);
    cout << res;
    return 0;
}

File: Python/ps_baekjoon/2157/main.py
import sys
from collections import defaultdict

readline, write = (sys.stdin.readline, sys.stdout.write)

n, m, k = [int(i) for i in readline().split()]
graph = [defaultdict(int) for _ in range(n + 1)]
dp = [[None] + [0] + [-float("inf")] * (n - 1)] + [[-float("inf")] * (n + 1) for _ in range(m - 1)]
for _ in range(k):
    a, b, c = [int(i) for i in readline().split()]
    if not a < b:
        continue
    graph[a][b] = max(graph[a][b], c)
for frm in range(1, n + 1):
    for to, cost in graph[frm].items():
        for cnt in range(1, m):
            dp[cnt][to] = max(dp[cnt][to], dp[cnt - 1][frm] + cost)
write(f"{max([dp[i][n] for i in range(1, m)])}")